src/firebase-app.ts

```diff
@@ -57,129 +59,80 @@ export interface FirebaseAccessToken {
-            this.cachedToken_ = token;
-            this.tokenListeners_.forEach((listener) => {
-              listener(token.accessToken);
-            });
-          }
-
-          // Establish a timeout to proactively refresh the token every minute starting at five
-          // minutes before it expires. Once a token refresh succeeds, no further retries are
-          // needed; if it fails, retry every minute until the token expires (resulting in a total
-          // of four retries: at 4, 3, 2, and 1 minutes).
-          let refreshTimeInSeconds = (result.expires_in - (5 * 60));
-          let numRetries = 4;
-
-          // In the rare cases the token is short-lived (that is, it expires in less than five
-          // minutes from when it was fetched), establish the timeout to refresh it after the
-          // current minute ends and update the number of retries that should be attempted before
-          // the token expires.
-          if (refreshTimeInSeconds <= 0) {
-            refreshTimeInSeconds = result.expires_in % 60;
-            numRetries = Math.floor(result.expires_in / 60) - 1;
-          }
-
-          // The token refresh timeout keeps the Node.js process alive, so only create it if this
-          // instance has not already been deleted.
-          if (numRetries && !this.isDeleted_) {
-            this.setTokenRefreshTimeout(refreshTimeInSeconds * 1000, numRetries);
-          }
-
-          return token;
-        })
-        .catch((error) => {
-          let errorMessage = (typeof error === 'string') ? error : error.message;
-
-          errorMessage = 'Credential implementation provided to initializeApp() via the ' +
-            '"credential" property failed to fetch a valid Google OAuth2 access token with the ' +
-            `following error: "${errorMessage}".`;
-
-          if (errorMessage.indexOf('invalid_grant') !== -1) {
-            errorMessage += ' There are two likely causes: (1) your server time is not properly ' +
-            'synced or (2) your certificate key file has been revoked. To solve (1), re-sync the ' +
-            'time on your server. To solve (2), make sure the key ID for your key file is still ' +
-            'present at https://console.firebase.google.com/iam-admin/serviceaccounts/project. If ' +
-            'not, generate a new key file at ' +
-            'https://console.firebase.google.com/project/_/settings/serviceaccounts/adminsdk.';
-          }
-
-          throw new FirebaseAppError(AppErrorCodes.INVALID_CREDENTIAL, errorMessage);
-        });
-
-      return this.cachedTokenPromise_;
+  public getToken(forceRefresh = false): Promise<FirebaseAccessToken> {
+    if (forceRefresh || this.shouldRefresh()) {
+      return this.refreshToken();
     }
+
+    return Promise.resolve(this.cachedToken_);
+  }
+
+  private refreshToken(): Promise<FirebaseAccessToken> {
+    return Promise.resolve(this.credential_.getAccessToken())
```

Comment:
I'm confused. isn't getAccesssToken() already returning a promise? Why do we need to resolve the promise instead of just chaining?

Reply:
I thought so too :)

But when I made the change, it actually caused a test failure here:
https://github.com/firebase/firebase-admin-node/blob/994fd43e36c0824deba176763010dc141e9bb0f4/test/unit/firebase-app.spec.ts#L677-L691

This is because the `credential` in this case returns a value which cannot be chained. I'm not sure how much we care about this sort of edge cases, but for now I've opted to keep it as is.